src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs

```diff
@@ -61,11 +61,31 @@ protected ClaimsPrincipal(SerializationInfo info, StreamingContext context)
         {
             ArgumentNullException.ThrowIfNull(identities);
 
-            foreach (ClaimsIdentity identity in identities)
+            // If the identities value is exactly a List<ClaimsIdentity>, special case it so that
+            // the enumerator allocation can be skipped. Doing this for List<ClaimsIdentity> is the 99%
+            // case because it is normally used on the _identities value, which is a List<ClaimsIdentity>.
+            if (identities.GetType() == typeof(List<ClaimsIdentity>))
```

Comment:
We could _probably_ get away with `is List<ClaimsIdentity>` (not the interface). Nothing on `List<T>` is virtual, including `GetEnumerator`. If we want to relax that, we can, but I erred on the side of maximum compatibility.

Reply:
I think, if you care, that the test for that would use

```C#
    internal class WhyNot : List<ClaimsIdentity>, IEnumerable<ClaimsIdentity>
    {
        public int GetEnumeratorCount { get; private set; }

        public new IEnumerator<ClaimsIdentity> GetEnumerator()
        {
            GetEnumeratorCount++;
            return base.GetEnumerator();
        }
    }
```

When treated as an IEnumerable it increments GetEnumeratorCount, when treated as a List it doesn't (and the `, IEnumerable<ClaimsIdentity>` is important to get the interface virtual slot overridden); so the test would be ensuring that it observes the appropriate call-count.

That could replace your current IList-but-not-List, or be in addition to it.